### DISCUSSION TITLE
Does multipoint,multiPolyline,multiPolygon  support？

### DISCUSSION THREAD

**Discussion by LemonLiTree** [google/s2geometry#314] [General]
I did not find support for multipoint, multiPolyline, multiPolygon, does the community have any considerations?

**Comment by jmr** [comment]
An `S2Polygon` is essentially a `MultiPolygon`.  For the others, you can maybe use `S2RegionUnion`, depending on what you're trying to do.  What is the problem you're trying to solve?

**Reply by LemonLiTree** [reply]
Can I understand that when there is only one loop in Polygon, it is Polygon, and when there are multiple loops, it is multiPolygon？

**Reply by jmr** [reply]
Loops can be nested with the opposite orientation to create holes. I think geojson plain polygons can do this. I'm not really sure what "MultiPolygon" you're asking about.

If loops don't contain each other, then it's the union of separate polygons.  See `s2polygon.h` and `s2loop.h`.

**Reply by LemonLiTree** [reply]
MultiPolygon in GEOS

**Comment by smcallis** [comment]
S2Shape can be all of those because it can contain multiple chains of a particular dimension, that gives you multipoint and multipolyline, and then S2Shape can represent both simple and complex (disjoint interior area) polygons.  Whether you consider a 2D S2hape to be one polygon or a group of multiple polygons is a matter of interpretation.

**Comment by eengle** [comment]
A Polygon has one shell, or outermost ring. A MultiPolygon has multiple
shells. But on the sphere, there isn't really a good sense of "outermost".
Consider a polygon formed by a ring at 10 degrees north and a ring at 10
degrees south, oriented so that the polygon contains the 20 degree tall
band of latitude centered on the equator. Unlike 2D geometry where you need
to know which ring is the outermost, it doesn't matter on the sphere --
the +10 ring can be the shell with a hole at -10 latitude, or vice-versa.

Similarly, S2 operations aren't specialized on the dimension of the
geometries, so there are no specific types for MultiLine and MultiPoint.
But you can create collections, either as a simple vector of S2Shape*, or
as an S2ShapeIndex if you want to operate on the collection as a whole.
This indexed collection type is another great strength of S2 over libraries
like GEOS, which only provide much simpler indexes of envelopes. This makes
a large difference in performance when you have to work with complex lines
or polygons, especially.

Kind regards,

Eric
